File: apps/desktop/src/types/provider.ts
export type ModelType = "llm" | "image" | "video" | "audio";
export type ProviderProtocol =
  | "openai_compat"
  | "dashscope"
  | "sora"
  | "openrouter_video"
  | "zhipu_video"
  | "siliconflow_video"
  | "elevenlabs"
  | "syncso";

export const PROTOCOL_LABELS: Record<ProviderProtocol, string> = {
  openai_compat: "OpenAI 兼容",
  dashscope: "阿里云百炼 / DashScope",
  sora: "OpenAI Sora",
  openrouter_video: "OpenRouter Video",
  zhipu_video: "智谱 CogVideoX",
  siliconflow_video: "SiliconFlow Video",
  elevenlabs: "ElevenLabs",
  syncso: "Sync.so",
};

export type CapabilityKey =
  | "text_to_image"
  | "image_to_image"
  | "reference_image"
  | "character_reference"
  | "text_to_video"
  | "image_to_video"
  | "video_to_video"
  | "first_frame"
  | "last_frame"
  | "first_last_frame"
  | "video_audio"
  | "text_to_speech"
  | "speech_to_text"
  | "tts_timestamps"
  | "voice_clone"
  | "voice_design"
  | "lip_sync";

export const CAPABILITY_LABELS: Record<CapabilityKey, string> = {
  text_to_image: "文生图",
  image_to_image: "图生图",
  reference_image: "参考图",
  character_reference: "角色参考",
  text_to_video: "文生视频",
  image_to_video: "图生视频",
  video_to_video: "视频生视频",
  first_frame: "首帧控制",
  last_frame: "尾帧控制",
  first_last_frame: "首尾帧控制",
  video_audio: "视频自带音效",
  text_to_speech: "文本转语音",
  speech_to_text: "语音转写",
  tts_timestamps: "语音时间戳",
  voice_clone: "声音复刻",
  voice_design: "声音设计",
  lip_sync: "对口型",
};

export const IMAGE_CAPABILITIES: CapabilityKey[] = [
  "text_to_image",
  "image_to_image",
  "reference_image",
  "character_reference",
];

export const VIDEO_CAPABILITIES: CapabilityKey[] = [
  "text_to_video",
  "image_to_video",
  "video_to_video",
  "first_frame",
  "last_frame",
  "first_last_frame",
  "video_audio",
];

export const AUDIO_CAPABILITIES: CapabilityKey[] = [
  "text_to_speech",
  "tts_timestamps",
  "voice_clone",
  "voice_design",
];

export interface Preset {
  key: string;
  name: string;
  base_url: string;
  needs_key: boolean;
  protocol: ProviderProtocol;
  discoverable: boolean;
}

export interface Provider {
  id: string;
  name: string;
  preset_key: string | null;
  protocol: ProviderProtocol;
  api_base_url: string;
  needs_key: boolean;
  enabled: boolean;
  has_api_key: boolean;
  model_count: number;
  created_at: string;
  updated_at: string;
}

export interface ProviderInput {
  name?: string;
  preset_key?: string;
  protocol?: ProviderProtocol;
  api_base_url?: string;
  needs_key?: boolean;
  api_key?: string;
  enabled?: boolean;
}

export interface Model {
  id: string;
  provider_id: string;
  provider_name: string;
  provider_preset_key?: string | null;
  provider_protocol?: ProviderProtocol;
  provider_base_url: string;
  provider_needs_key: boolean;
  provider_has_api_key: boolean;
  model_id: string;
  model_type: ModelType;
  capabilities: CapabilityKey[];
  capability_source: "auto" | "manual";
  enabled: boolean;
  is_default_image: boolean;
  is_default_video: boolean;
  created_at: string;
  updated_at: string;
}

export interface ModelInput {
  provider_id: string;
  model_id: string;
  model_type: ModelType;
  enabled?: boolean;
}

export interface BuiltinModel {
  id: string;
  type: ModelType;
  capabilities: CapabilityKey[];
}

export interface ProviderCheck {
  label: string;
  status: "ok" | "fail" | "skipped";
  detail: string;
}

export interface ModelCheck {
  model_id: string;
  ok: boolean;
  detail: string;
}

export interface ProviderTestResult {
  provider_id: string;
  ok: boolean;
  checks: ProviderCheck[];
  model_checks: ModelCheck[];
  tested_at: string;
}

File: apps/backend/app/schemas/provider.py
"""Provider / Model API schemas。

响应永不包含 API Key，只提供 has_api_key 状态。
模型显示名 = model_id（不允许用户取名）。
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

ModelType = Literal["llm", "image", "video", "audio"]


class ProviderCreate(BaseModel):
    name: str | None = Field(default=None, max_length=100)
    preset_key: str | None = None
    protocol: str | None = None
    api_base_url: str = Field(default="", max_length=500)
    needs_key: bool = True
    api_key: str | None = Field(default=None, max_length=1000)


class ProviderUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    protocol: str | None = None
    api_base_url: str | None = Field(default=None, max_length=500)
    needs_key: bool | None = None
    enabled: bool | None = None
    api_key: str | None = Field(default=None, max_length=1000)


class ProviderOut(BaseModel):
    id: str
    name: str
    preset_key: str | None = None
    protocol: str = "openai_compat"
    api_base_url: str = ""
    needs_key: bool = True
    enabled: bool = True
    has_api_key: bool = False
    model_count: int = 0
    created_at: datetime
    updated_at: datetime


class PresetOut(BaseModel):
    key: str
    name: str
    base_url: str
    needs_key: bool
    protocol: str = "openai_compat"
    discoverable: bool = True


class ModelCreate(BaseModel):
    provider_id: str
    model_id: str = Field(min_length=1, max_length=200)
    model_type: ModelType = "llm"
    enabled: bool = True
    is_default_image: bool = False
    is_default_video: bool = False


class ModelUpdate(BaseModel):
    model_type: ModelType | None = None
    enabled: bool | None = None
    is_default_image: bool | None = None
    is_default_video: bool | None = None


class ModelCapabilityUpdate(BaseModel):
    """能力覆盖：source=manual 使用 capabilities；source=auto 重新按规则推断。"""

    capabilities: list[str] = Field(default_factory=list)
    source: Literal["auto", "manual"] = "manual"


class DefaultRequest(BaseModel):
    model_type: Literal["image", "video"] | None = None


class BulkModelsRequest(BaseModel):
    model_ids: list[str] = Field(min_length=1)


class BuiltinModelOut(BaseModel):
    id: str
    type: ModelType
    capabilities: list[str] = Field(default_factory=list)


class ModelOut(BaseModel):
    id: str
    provider_id: str
    provider_name: str
    provider_preset_key: str | None = None
    provider_protocol: str = "openai_compat"
    provider_enabled: bool = True
    provider_base_url: str = ""
    provider_needs_key: bool = True
    provider_has_api_key: bool = False
    model_id: str
    model_type: ModelType
    capabilities: list[str] = Field(default_factory=list)
    capability_source: Literal["auto", "manual"] = "auto"
    enabled: bool = True
    is_default_image: bool = False
    is_default_video: bool = False
    created_at: datetime
    updated_at: datetime


class ProviderCheckOut(BaseModel):
    label: str
    status: Literal["ok", "fail", "skipped"]
    detail: str = ""


class ModelCheckOut(BaseModel):
    model_id: str
    ok: bool
    detail: str = ""


class ProviderTestOut(BaseModel):
    provider_id: str
    ok: bool
    checks: list[ProviderCheckOut] = Field(default_factory=list)
    model_checks: list[ModelCheckOut] = Field(default_factory=list)
    tested_at: datetime
